Inconsistent wiring-in of promoted/singled names from the Prelude

This code best illustrates the problem:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# OPTIONS_GHC -Wall #-}
module Bug where

import Data.Singletons.Prelude.Bool (SBool(..), FalseSym0, TrueSym0)
import Data.Singletons.Prelude.Eq (PEq(..))
import Data.Singletons.Prelude.Eq (SEq(..))
import Data.Singletons.TH (singletons)

$(singletons [d| data T = MkT deriving Eq |])
```

If you compile this, you'll actually get a warning:

```
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )

Bug.hs:12:1: warning: [-Wunused-imports]
    The import of ‘Data.Singletons.Prelude.Eq’ is redundant
      except perhaps to import instances from ‘Data.Singletons.Prelude.Eq’
    To import instances alone, use: import Data.Singletons.Prelude.Eq()
   |
12 | import Data.Singletons.Prelude.Eq (SEq(..))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

This is rather surprising. After all, `singletons [d| ... deriving Eq |]` should generate an `SEq` instance, so why would importing `SEq` be considered redundant? To make things even more confusing, if you comment out the `PEq` import, that will result in an error:

```
Bug.hs:14:3: error:
    Not in scope: type constructor or class ‘PEq’
    Perhaps you meant ‘Eq’ (imported from Prelude)
    Perhaps you want to add ‘PEq’ to the import list in the import of
    ‘Data.Singletons.TH’ (Bug.hs:12:1-38).
   |
14 | $(singletons [d| data T = MkT deriving Eq |])
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

What's going on here? To reveal the full story, let's comment out the `SEq` import and compile with `-ddump-splices`:

```
Bug.hs:16:3-44: Splicing declarations
    singletons
      [d| data T_akCn
            = MkT_akCo
            deriving Eq |]
  ======>
    ...
    instance PEq T_al3x where ...
    instance Data.Singletons.Prelude.Eq.SEq T_al3x where ...
```

Notice how the generated `PEq` instance uses the name "`PEq`" unqualified, which demands that it be imported from elsewhere in order to be in scope. The generated `SEq` instance, on the other hand, uses "`Data.Singletons.Prelude.Eq.SEq`", which is fully qualified. As a result, GHC always figures out where `SEq` lives, even if it is not explicitly imported.

I find this state of affairs confusing, and I believe we should pick a consistent policy for all autogenerated names. My vote would be to generate all such names unqualified (like `PEq` above), although I could be convinced otherwise if others feel strongly.

--- a/src/Data/Singletons/Deriving/Enum.hs
+++ b/src/Data/Singletons/Deriving/Enum.hs
@@ -53,11 +53,8 @@ mkEnumInstance mb_ctxt ty (DataDecl _ _ cons) = do
                                                         (DLitE (IntegerL i)))
                                      [0..] cons)
   return (InstDecl { id_cxt     = fromMaybe [] mb_ctxt
-                   , id_name    = singletonsEnumName
-                      -- need to use singletons's Enum class to get the types
-                      -- to use Nat instead of Int
-
+                   , id_name    = enumName
                    , id_arg_tys = [ty]
                    , id_sigs    = mempty
-                   , id_meths   = [ (singletonsToEnumName, to_enum)
-                                  , (singletonsFromEnumName, from_enum) ] })
+                   , id_meths   = [ (toEnumName, to_enum)
+                                  , (fromEnumName, from_enum) ] })

--- a/src/Data/Singletons/Names.hs
+++ b/src/Data/Singletons/Names.hs
@@ -18,6 +18,7 @@ import Language.Haskell.TH.Desugar
 import GHC.TypeLits ( Nat, Symbol )
 import GHC.Exts ( Constraint )
 import GHC.Show ( showCommaSpace, showSpace )
+import Data.String (fromString)
 import Data.Type.Equality ( TestEquality(..) )
 import Data.Type.Coercion ( TestCoercion(..) )
 import Data.Typeable ( TypeRep )
@@ -25,29 +26,57 @@ import Data.Singletons.Util
 import Control.Applicative
 import Control.Monad
 
-boolName, andName, tyEqName, compareName, minBoundName,
+{-
+Note [Wired-in Names]
+~~~~~~~~~~~~~~~~~~~~~
+The list of Names below contains everything that the Template Haskell machinery
+needs to have special knowledge of. These names can be broadly categorized into
+two groups:
+
+1. Names of basic singleton definitions (Sing, SingKind, etc.). These are
+   spliced directly into TH-generated code.
+2. Names of definitions from the Prelude. These are not spliced into
+   TH-generated code, but are instead used as the namesakes for promoted and
+   singled definitions. For example, the TH machinery must be aware of the Name
+   `fromInteger` so that it can promote and single the expression `42` to
+   `FromInteger 42` and `sFromInteger (sing @42)`, respectively.
+
+Note that we deliberately do not wire in promoted or singled Names, such as
+FromInteger or sFromInteger, for two reasons:
+
+a. We want all promoted and singled names to go through the naming options in
+   D.S.TH.Options. Splicing the name FromInteger directly into TH-generated
+   code, for instance, would prevent users from overriding the default options
+   in order to promote `fromInteger` to something else (e.g.,
+   MyCustomFromInteger).
+b. Wired in names live in particular modules, so if we were to wire in the name
+   FromInteger, it would come from D.S.Prelude.Num. This would effectively
+   prevent anyone from defining their own version of FromInteger and
+   piggybacking on top of the TH machinery to generate it, however. As a
+   result, we generate the name FromInteger completely unqualified so that
+   it picks up whichever version of FromInteger is in scope.
+-}
+
+boolName, andName, compareName, minBoundName,
   maxBoundName, repName,
   nilName, consName, listName, tyFunArrowName,
   applyName, applyTyConName, applyTyConAux1Name,
   natName, symbolName, typeRepName, stringName,
   eqName, ordName, boundedName, orderingName,
   singFamilyName, singIName, singMethName, demoteName, withSingIName,
-  singKindClassName, sEqClassName, sEqMethName, sconsName, snilName, strueName,
-  sIfName,
-  someSingTypeName, someSingDataName,
-  sListName, sDecideClassName, sDecideMethName,
+  singKindClassName, someSingTypeName, someSingDataName,
+  sDecideClassName, sDecideMethName,
   testEqualityClassName, testEqualityMethName, decideEqualityName,
   testCoercionClassName, testCoercionMethName, decideCoercionName,
   provedName, disprovedName, reflName, toSingName, fromSingName,
   equalityName, applySingName, suppressClassName, suppressMethodName,
-  thenCmpName,
-  sameKindName, tyFromIntegerName, tyNegateName, sFromIntegerName,
-  sNegateName, errorName, foldlName, cmpEQName, cmpLTName, cmpGTName,
-  singletonsToEnumName, singletonsFromEnumName, enumName, singletonsEnumName,
+  thenCmpName, sameKindName, fromIntegerName, negateName,
+  errorName, foldlName, cmpEQName, cmpLTName, cmpGTName,
+  toEnumName, fromEnumName, enumName,
   equalsName, constraintName,
   showName, showSName, showCharName, showCommaSpaceName, showParenName, showsPrecName,
   showSpaceName, showStringName, showSingName, showSing'Name,
-  composeName, gtName, tyFromStringName, sFromStringName,
+  composeName, gtName, fromStringName,
   foldableName, foldMapName, memptyName, mappendName, foldrName,
   functorName, fmapName, replaceName,
   traversableName, traverseName, pureName, apName, liftA2Name :: Name
@@ -56,7 +85,6 @@ andName = '(&&)
 compareName = 'compare
 minBoundName = 'minBound
 maxBoundName = 'maxBound
-tyEqName = mk_name_tc "Data.Singletons.Prelude.Eq" "=="
 repName = mkName "Rep"   -- this is actually defined in client code!
 nilName = '[]
 consName = '(:)
@@ -81,15 +109,8 @@ fromSingName = 'fromSing
 demoteName = ''Demote
 withSingIName = 'withSingI
 singKindClassName = ''SingKind
-sEqClassName = mk_name_tc "Data.Singletons.Prelude.Eq" "SEq"
-sEqMethName = mk_name_v "Data.Singletons.Prelude.Eq" "%=="
-sIfName = mk_name_v "Data.Singletons.Prelude.Bool" "sIf"
-sconsName = mk_name_d "Data.Singletons.Prelude.Instances" "SCons"
-snilName = mk_name_d "Data.Singletons.Prelude.Instances" "SNil"
-strueName = mk_name_d "Data.Singletons.Prelude.Instances" "STrue"
 someSingTypeName = ''SomeSing
 someSingDataName = 'SomeSing
-sListName = mk_name_tc "Data.Singletons.Prelude.Instances" "SList"
 sDecideClassName = ''SDecide
 sDecideMethName = '(%~)
 testEqualityClassName = ''TestEquality
@@ -105,21 +126,18 @@ equalityName = ''(~)
 applySingName = 'applySing
 suppressClassName = ''SuppressUnusedWarnings
 suppressMethodName = 'suppressUnusedWarnings
-thenCmpName = mk_name_v "Data.Singletons.Prelude.Ord" "thenCmp"
+thenCmpName = 'thenCmp
 sameKindName = ''SameKind
-tyFromIntegerName = mk_name_tc "Data.Singletons.Prelude.Num" "FromInteger"
-tyNegateName = mk_name_tc "Data.Singletons.Prelude.Num" "Negate"
-sFromIntegerName = mk_name_v "Data.Singletons.Prelude.Num" "sFromInteger"
-sNegateName = mk_name_v "Data.Singletons.Prelude.Num" "sNegate"
+fromIntegerName = 'fromInteger
+negateName = 'negate
 errorName = 'error
 foldlName = 'foldl
 cmpEQName = 'EQ
 cmpLTName = 'LT
 cmpGTName = 'GT
-singletonsToEnumName = mk_name_v "Data.Singletons.Prelude.Enum" "toEnum"
-singletonsFromEnumName = mk_name_v "Data.Singletons.Prelude.Enum" "fromEnum"
+toEnumName = 'toEnum
+fromEnumName = 'fromEnum
 enumName = ''Enum
-singletonsEnumName = mk_name_tc "Data.Singletons.Prelude.Enum" "Enum"
 equalsName = '(==)
 constraintName = ''Constraint
 showName = ''Show
@@ -134,8 +152,7 @@ showSing'Name = mk_name_tc "Data.Singletons.ShowSing" "ShowSing'"
 composeName = '(.)
 gtName = '(>)
 showCommaSpaceName = 'showCommaSpace
-tyFromStringName = mk_name_tc "Data.Singletons.Prelude.IsString" "FromString"
-sFromStringName = mk_name_v "Data.Singletons.Prelude.IsString" "sFromString"
+fromStringName = 'fromString
 foldableName = ''Foldable
 foldMapName = 'foldMap
 memptyName = 'mempty
@@ -155,20 +172,6 @@ singPkg = $( (LitE . StringL . loc_package) `liftM` location )
 
 mk_name_tc :: String -> String -> Name
 mk_name_tc = mkNameG_tc singPkg
-
-mk_name_d :: String -> String -> Name
-mk_name_d = mkNameG_d singPkg
-
-mk_name_v :: String -> String -> Name
-mk_name_v = mkNameG_v singPkg
-
-mkTupleTypeName :: Int -> Name
-mkTupleTypeName n = mk_name_tc "Data.Singletons.Prelude.Instances" $
-                    "STuple" ++ (show n)
-
-mkTupleDataName :: Int -> Name
-mkTupleDataName n = mk_name_d "Data.Singletons.Prelude.Instances" $
-                    "STuple" ++ (show n)
 
 mkTyName :: Quasi q => Name -> q Name
 mkTyName tmName = do

--- a/src/Data/Singletons/Prelude/Ord.hs
+++ b/src/Data/Singletons/Prelude/Ord.hs
@@ -50,6 +50,7 @@ import Data.Singletons.Single
 import Data.Singletons.Prelude.Eq
 import Data.Singletons.Prelude.Instances
 import Data.Singletons.Util
+import Language.Haskell.TH.Syntax (thenCmp)
 
 $(singletonsOnly [d|
   class  (Eq a) => Ord a  where
@@ -88,6 +89,11 @@ $(singletonsOnly [d|
   -- >   ... sortBy (comparing fst) ...
   comparing :: (Ord a) => (b -> a) -> b -> b -> Ordering
   comparing p x y = compare (p x) (p y)
+
+  thenCmp :: Ordering -> Ordering -> Ordering
+  thenCmp EQ x = x
+  thenCmp LT _ = LT
+  thenCmp GT _ = GT
   |])
 
 $(genSingletons [''Down])
@@ -97,13 +103,6 @@ $(singletonsOnly [d|
 
   instance Ord a => Ord (Down a) where
       compare (Down x) (Down y) = y `compare` x
-  |])
-
-$(singletons [d|
-  thenCmp :: Ordering -> Ordering -> Ordering
-  thenCmp EQ x = x
-  thenCmp LT _ = LT
-  thenCmp GT _ = GT
   |])
 
 $(singOrdInstances basicTypes)

--- a/src/Data/Singletons/Prelude/Proxy.hs
+++ b/src/Data/Singletons/Prelude/Proxy.hs
@@ -47,6 +47,7 @@ import Data.Singletons.Prelude.Eq
 import Data.Singletons.Prelude.Instances
 import Data.Singletons.Prelude.Monad.Internal
 import Data.Singletons.Prelude.Monoid
+import Data.Singletons.Prelude.Num
 import Data.Singletons.Prelude.Ord
 import Data.Singletons.Prelude.Semigroup.Internal
 import Data.Singletons.Prelude.Show

--- a/src/Data/Singletons/Prelude/Show.hs
+++ b/src/Data/Singletons/Prelude/Show.hs
@@ -60,6 +60,7 @@ import           Data.Singletons.Internal
 import           Data.Singletons.Prelude.Base
 import           Data.Singletons.Prelude.Instances
 import           Data.Singletons.Prelude.List.Internal
+import           Data.Singletons.Prelude.Num
 import           Data.Singletons.Prelude.Ord
 import           Data.Singletons.Prelude.Semigroup.Internal
 import           Data.Singletons.Promote

--- a/src/Data/Singletons/Promote.hs
+++ b/src/Data/Singletons/Promote.hs
@@ -1106,16 +1106,21 @@ promoteExp (DSigE exp ty) = do
   return (DSigT exp' ty', ADSigE exp' ann_exp ty')
 promoteExp e@(DStaticE _) = fail ("Static expressions cannot be promoted: " ++ show e)
 
-promoteLitExp :: Quasi q => Lit -> q DType
-promoteLitExp (IntegerL n)
-  | n >= 0    = return $ (DConT tyFromIntegerName `DAppT` DLitT (NumTyLit n))
-  | otherwise = return $ (DConT tyNegateName `DAppT`
-                          (DConT tyFromIntegerName `DAppT` DLitT (NumTyLit (-n))))
+promoteLitExp :: OptionsMonad q => Lit -> q DType
+promoteLitExp (IntegerL n) = do
+  opts <- getOptions
+  let tyFromIntegerName = promotedValueName opts fromIntegerName Nothing
+      tyNegateName      = promotedValueName opts negateName      Nothing
+  if n >= 0
+     then return $ (DConT tyFromIntegerName `DAppT` DLitT (NumTyLit n))
+     else return $ (DConT tyNegateName `DAppT`
+                    (DConT tyFromIntegerName `DAppT` DLitT (NumTyLit (-n))))
 promoteLitExp (StringL str) = do
+  opts <- getOptions
   let prom_str_lit = DLitT (StrTyLit str)
   os_enabled <- qIsExtEnabled LangExt.OverloadedStrings
   pure $ if os_enabled
-         then DConT tyFromStringName `DAppT` prom_str_lit
+         then DConT (promotedValueName opts fromStringName Nothing) `DAppT` prom_str_lit
          else prom_str_lit
 promoteLitExp lit =
   fail ("Only string and natural number literals can be promoted: " ++ show lit)

--- a/src/Data/Singletons/Single.hs
+++ b/src/Data/Singletons/Single.hs
@@ -989,20 +989,20 @@ singDerivedEqDecs (DerivedDecl { ded_mb_cxt     = mb_ctxt
   -- When we single the context, it will become (SEq a). But we do *not* want
   -- this for the SDecide instance! The simplest solution is to simply replace
   -- all occurrences of SEq with SDecide in the context.
-  let mb_sctxtDecide = fmap (map sEqToSDecide) mb_sctxt
+  mb_sctxtDecide <- traverse (traverse sEqToSDecide) mb_sctxt
   sDecideInst <- mkDecideInstance mb_sctxtDecide kind cons scons
   testInsts <- traverse (mkTestInstance mb_sctxtDecide kind ty_tycon cons)
                         [TestEquality, TestCoercion]
   return (sDecideInst:testInsts)
 
 -- Walk a DPred, replacing all occurrences of SEq with SDecide.
-sEqToSDecide :: DPred -> DPred
-sEqToSDecide = modifyConNameDType $ \n ->
-  -- Why don't we directly compare n to sEqClassName? Because n is almost certainly
-  -- produced from a call to singClassName, which uses unqualified Names. Ugh.
-  if nameBase n == nameBase sEqClassName
-     then sDecideClassName
-     else n
+sEqToSDecide :: OptionsMonad q => DPred -> q DPred
+sEqToSDecide p = do
+  opts <- getOptions
+  pure $ modifyConNameDType (\n ->
+         if n == singledClassName opts eqName
+            then sDecideClassName
+            else n) p
 
 -- See Note [DerivedDecl] in Data.Singletons.Syntax
 singDerivedShowDecs :: DerivedShowDecl -> SgM [DDec]
@@ -1056,19 +1056,22 @@ singMatch res_ki (ADMatch var_proms pat exp) = do
   return $ DMatch sPat $ mkSigPaCaseE sigPaExpsSigs sExp
 
 singLit :: Lit -> SgM DExp
-singLit (IntegerL n)
-  | n >= 0    = return $
-                DVarE sFromIntegerName `DAppE`
-                (DVarE singMethName `DSigE`
-                 (singFamily `DAppT` DLitT (NumTyLit n)))
-  | otherwise = do sLit <- singLit (IntegerL (-n))
-                   return $ DVarE sNegateName `DAppE` sLit
+singLit (IntegerL n) = do
+  opts <- getOptions
+  if n >= 0
+     then return $
+          DVarE (singledValueName opts fromIntegerName) `DAppE`
+          (DVarE singMethName `DSigE`
+           (singFamily `DAppT` DLitT (NumTyLit n)))
+     else do sLit <- singLit (IntegerL (-n))
+             return $ DVarE (singledValueName opts negateName) `DAppE` sLit
 singLit (StringL str) = do
+  opts <- getOptions
   let sing_str_lit = DVarE singMethName `DSigE`
                      (singFamily `DAppT` DLitT (StrTyLit str))
   os_enabled <- qIsExtEnabled LangExt.OverloadedStrings
   pure $ if os_enabled
-         then DVarE sFromStringName `DAppE` sing_str_lit
+         then DVarE (singledValueName opts fromStringName) `DAppE` sing_str_lit
          else sing_str_lit
 singLit lit =
   fail ("Only string and natural number literals can be singled: " ++ show lit)

--- a/src/Data/Singletons/TH.hs
+++ b/src/Data/Singletons/TH.hs
@@ -48,7 +48,7 @@ module Data.Singletons.TH (
   cases, sCases,
 
   -- * Basic singleton definitions
-  SBool(..), STuple0(..), STuple2(..), STuple3(..), STuple4(..),
+  SList(..), SBool(..), STuple0(..), STuple2(..), STuple3(..), STuple4(..),
   STuple5(..), STuple6(..), STuple7(..), SOrdering(..),
   module Data.Singletons,
 
@@ -61,9 +61,10 @@ module Data.Singletons.TH (
   SDecide(..), (:~:)(..), Void, Refuted, Decision(..),
   PBounded(..), SBounded(..),
   PEnum(FromEnum, ToEnum), SEnum(sFromEnum, sToEnum),
-  PShow(..), SShow(..),
+  PShow(..), SShow(..), PIsString(..), SIsString(..),
   ShowString, sShowString, ShowParen, sShowParen, ShowSpace, sShowSpace,
   ShowChar, sShowChar, ShowCommaSpace, sShowCommaSpace,
+  FromInteger, sFromInteger, Negate, sNegate,
   PFunctor(..), SFunctor(..),
   PFoldable(..), SFoldable(..), PMonoid(..), SMonoid(..),
   PTraversable(..), STraversable(..), PApplicative(..), SApplicative(..),
@@ -93,6 +94,9 @@ module Data.Singletons.TH (
   ShowSpaceSym0, ShowSpaceSym1,
   ShowCharSym0, ShowCharSym1, ShowCharSym2,
   ShowCommaSpaceSym0, ShowCommaSpaceSym1,
+  FromIntegerSym0, FromIntegerSym1,
+  NegateSym0, NegateSym1,
+  FromStringSym0, FromStringSym1,
   FmapSym0, FmapSym1, FmapSym2,
   type (<$@#@$),  type (<$@#@$$),  type (<$@#@$$$),
   FoldMapSym0, FoldMapSym1, FoldMapSym2,
@@ -123,7 +127,9 @@ import Data.Singletons.Prelude.Enum
 import Data.Singletons.Prelude.Eq
 import Data.Singletons.Prelude.Foldable
 import Data.Singletons.Prelude.Functor hiding (Void)
+import Data.Singletons.Prelude.IsString
 import Data.Singletons.Prelude.Monoid
+import Data.Singletons.Prelude.Num
 import Data.Singletons.Prelude.Ord
 import Data.Singletons.Prelude.Show
 import Data.Singletons.Prelude.Traversable

--- a/src/Data/Singletons/TH/Options.hs
+++ b/src/Data/Singletons/TH/Options.hs
@@ -45,11 +45,7 @@ import Control.Monad.RWS (RWST)
 import Control.Monad.State (StateT)
 import Control.Monad.Trans.Class (MonadTrans(..))
 import Control.Monad.Writer (WriterT)
-import Data.Singletons.Names ( consName, listName, nilName
-                             , mk_name_tc, mkTupleDataName, mkTupleTypeName
-                             , sconsName, sListName, snilName
-                             , splitUnderscores
-                             )
+import Data.Singletons.Names (consName, listName, nilName, splitUnderscores)
 import Data.Singletons.Util
 import Language.Haskell.TH.Desugar
 import Language.Haskell.TH.Syntax hiding (Lift(..))
@@ -223,8 +219,7 @@ promoteTySym name sat
        -- treat unboxed tuples like tuples
     | Just degree <- tupleNameDegree_maybe name <|>
                      unboxedTupleNameDegree_maybe name
-    = mk_name_tc "Data.Singletons.Prelude.Instances" $
-                 "Tuple" ++ show degree ++ "Sym" ++ (show sat)
+    = mkName $ "Tuple" ++ show degree ++ "Sym" ++ show sat
 
     | otherwise
     = default_case name
@@ -244,18 +239,21 @@ promoteClassName = prefixName "P" "#"
 
 singDataConName :: Name -> Name
 singDataConName nm
-  | nm == nilName                                  = snilName
-  | nm == consName                                 = sconsName
-  | Just degree <- tupleNameDegree_maybe nm        = mkTupleDataName degree
-  | Just degree <- unboxedTupleNameDegree_maybe nm = mkTupleDataName degree
+  | nm == nilName                                  = mkName "SNil"
+  | nm == consName                                 = mkName "SCons"
+  | Just degree <- tupleNameDegree_maybe nm        = mkTupleName degree
+  | Just degree <- unboxedTupleNameDegree_maybe nm = mkTupleName degree
   | otherwise                                      = prefixConName "S" "%" nm
 
 singTyConName :: Name -> Name
 singTyConName name
-  | name == listName                                 = sListName
-  | Just degree <- tupleNameDegree_maybe name        = mkTupleTypeName degree
-  | Just degree <- unboxedTupleNameDegree_maybe name = mkTupleTypeName degree
+  | name == listName                                 = mkName "SList"
+  | Just degree <- tupleNameDegree_maybe name        = mkTupleName degree
+  | Just degree <- unboxedTupleNameDegree_maybe name = mkTupleName degree
   | otherwise                                        = prefixName "S" "%" name
+
+mkTupleName :: Int -> Name
+mkTupleName n = mkName $ "STuple" ++ show n
 
 singClassName :: Name -> Name
 singClassName = singTyConName

--- a/tests/SingletonsTestSuite.hs
+++ b/tests/SingletonsTestSuite.hs
@@ -135,6 +135,7 @@ tests =
     , afterSingletonsNat .
       compileAndDumpStdTest "T445"
     , compileAndDumpStdTest "T453"
+    , compileAndDumpStdTest "NegativeLiterals"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/tests/compile-and-dump/Singletons/EnumDeriving.golden
+++ b/tests/compile-and-dump/Singletons/EnumDeriving.golden
@@ -29,13 +29,13 @@ Singletons/EnumDeriving.hs:(0,0)-(0,0): Splicing declarations
       Case_0123456789876543210 n 'False = Apply ErrorSym0 "toEnum: bad argument"
     type family Case_0123456789876543210 n t where
       Case_0123456789876543210 n 'True = BazSym0
-      Case_0123456789876543210 n 'False = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 2))
+      Case_0123456789876543210 n 'False = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 2))
     type family Case_0123456789876543210 n t where
       Case_0123456789876543210 n 'True = BarSym0
-      Case_0123456789876543210 n 'False = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 1))
+      Case_0123456789876543210 n 'False = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 1))
     type ToEnum_0123456789876543210 :: GHC.Types.Nat -> Foo
     type family ToEnum_0123456789876543210 a where
-      ToEnum_0123456789876543210 n = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 0))
+      ToEnum_0123456789876543210 n = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 0))
     type ToEnum_0123456789876543210Sym0 :: (~>) GHC.Types.Nat Foo
     data ToEnum_0123456789876543210Sym0 a0123456789876543210
       where
@@ -50,9 +50,9 @@ Singletons/EnumDeriving.hs:(0,0)-(0,0): Splicing declarations
       ToEnum_0123456789876543210Sym1 a0123456789876543210 = ToEnum_0123456789876543210 a0123456789876543210
     type FromEnum_0123456789876543210 :: Foo -> GHC.Types.Nat
     type family FromEnum_0123456789876543210 a where
-      FromEnum_0123456789876543210 Bar = Data.Singletons.Prelude.Num.FromInteger 0
-      FromEnum_0123456789876543210 Baz = Data.Singletons.Prelude.Num.FromInteger 1
-      FromEnum_0123456789876543210 Bum = Data.Singletons.Prelude.Num.FromInteger 2
+      FromEnum_0123456789876543210 Bar = FromInteger 0
+      FromEnum_0123456789876543210 Baz = FromInteger 1
+      FromEnum_0123456789876543210 Bum = FromInteger 2
     type FromEnum_0123456789876543210Sym0 :: (~>) Foo GHC.Types.Nat
     data FromEnum_0123456789876543210Sym0 a0123456789876543210
       where
@@ -106,35 +106,32 @@ Singletons/EnumDeriving.hs:(0,0)-(0,0): Splicing declarations
                                                                      -> GHC.Types.Type) t)
       sToEnum (sN :: Sing n)
         = (id
-             @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 0)))))
+             @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 0)))))
             (case
                  (applySing ((applySing ((singFun2 @(==@#@$)) (%==))) sN))
-                   (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 0))
+                   (sFromInteger (sing :: Sing 0))
              of
                STrue -> SBar
                SFalse
                  -> (id
-                       @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 1)))))
+                       @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 1)))))
                       (case
                            (applySing ((applySing ((singFun2 @(==@#@$)) (%==))) sN))
-                             (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 1))
+                             (sFromInteger (sing :: Sing 1))
                        of
                          STrue -> SBaz
                          SFalse
                            -> (id
-                                 @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 2)))))
+                                 @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 2)))))
                                 (case
                                      (applySing ((applySing ((singFun2 @(==@#@$)) (%==))) sN))
-                                       (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 2))
+                                       (sFromInteger (sing :: Sing 2))
                                  of
                                    STrue -> SBum
                                    SFalse -> sError (sing :: Sing "toEnum: bad argument"))))
-      sFromEnum SBar
-        = Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 0)
-      sFromEnum SBaz
-        = Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 1)
-      sFromEnum SBum
-        = Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 2)
+      sFromEnum SBar = sFromInteger (sing :: Sing 0)
+      sFromEnum SBaz = sFromInteger (sing :: Sing 1)
+      sFromEnum SBum = sFromInteger (sing :: Sing 2)
     instance SingI Bar where
       sing = SBar
     instance SingI Baz where
@@ -153,10 +150,10 @@ Singletons/EnumDeriving.hs:0:0:: Splicing declarations
       Case_0123456789876543210 n 'False = Apply ErrorSym0 "toEnum: bad argument"
     type family Case_0123456789876543210 n t where
       Case_0123456789876543210 n 'True = Q1Sym0
-      Case_0123456789876543210 n 'False = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 1))
+      Case_0123456789876543210 n 'False = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 1))
     type ToEnum_0123456789876543210 :: GHC.Types.Nat -> Quux
     type family ToEnum_0123456789876543210 a where
-      ToEnum_0123456789876543210 n = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 0))
+      ToEnum_0123456789876543210 n = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 0))
     type ToEnum_0123456789876543210Sym0 :: (~>) GHC.Types.Nat Quux
     data ToEnum_0123456789876543210Sym0 a0123456789876543210
       where
@@ -171,8 +168,8 @@ Singletons/EnumDeriving.hs:0:0:: Splicing declarations
       ToEnum_0123456789876543210Sym1 a0123456789876543210 = ToEnum_0123456789876543210 a0123456789876543210
     type FromEnum_0123456789876543210 :: Quux -> GHC.Types.Nat
     type family FromEnum_0123456789876543210 a where
-      FromEnum_0123456789876543210 'Q1 = Data.Singletons.Prelude.Num.FromInteger 0
-      FromEnum_0123456789876543210 'Q2 = Data.Singletons.Prelude.Num.FromInteger 1
+      FromEnum_0123456789876543210 'Q1 = FromInteger 0
+      FromEnum_0123456789876543210 'Q2 = FromInteger 1
     type FromEnum_0123456789876543210Sym0 :: (~>) Quux GHC.Types.Nat
     data FromEnum_0123456789876543210Sym0 a0123456789876543210
       where
@@ -201,22 +198,20 @@ Singletons/EnumDeriving.hs:0:0:: Splicing declarations
                                                                      -> GHC.Types.Type) t)
       sToEnum (sN :: Sing n)
         = (id
-             @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 0)))))
+             @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 0)))))
             (case
                  (applySing ((applySing ((singFun2 @(==@#@$)) (%==))) sN))
-                   (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 0))
+                   (sFromInteger (sing :: Sing 0))
              of
                STrue -> SQ1
                SFalse
                  -> (id
-                       @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 1)))))
+                       @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 1)))))
                       (case
                            (applySing ((applySing ((singFun2 @(==@#@$)) (%==))) sN))
-                             (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 1))
+                             (sFromInteger (sing :: Sing 1))
                        of
                          STrue -> SQ2
                          SFalse -> sError (sing :: Sing "toEnum: bad argument")))
-      sFromEnum SQ1
-        = Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 0)
-      sFromEnum SQ2
-        = Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 1)
+      sFromEnum SQ1 = sFromInteger (sing :: Sing 0)
+      sFromEnum SQ2 = sFromInteger (sing :: Sing 1)

--- a/tests/compile-and-dump/Singletons/Maybe.golden
+++ b/tests/compile-and-dump/Singletons/Maybe.golden
@@ -55,7 +55,7 @@ Singletons/Maybe.hs:(0,0)-(0,0): Splicing declarations
                                           -> Maybe a -> GHC.Types.Symbol -> GHC.Types.Symbol
     type family ShowsPrec_0123456789876543210 a a a where
       ShowsPrec_0123456789876543210 _ Nothing a_0123456789876543210 = Apply (Apply ShowStringSym0 "Nothing") a_0123456789876543210
-      ShowsPrec_0123456789876543210 p_0123456789876543210 (Just arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (Data.Singletons.Prelude.Num.FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "Just ")) (Apply (Apply ShowsPrecSym0 (Data.Singletons.Prelude.Num.FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
+      ShowsPrec_0123456789876543210 p_0123456789876543210 (Just arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "Just ")) (Apply (Apply ShowsPrecSym0 (FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
     type ShowsPrec_0123456789876543210Sym0 :: (~>) GHC.Types.Nat ((~>) (Maybe a) ((~>) GHC.Types.Symbol GHC.Types.Symbol))
     data ShowsPrec_0123456789876543210Sym0 a0123456789876543210
       where
@@ -147,14 +147,14 @@ Singletons/Maybe.hs:(0,0)-(0,0): Splicing declarations
                  ((applySing ((singFun3 @ShowParenSym0) sShowParen))
                     ((applySing
                         ((applySing ((singFun2 @(>@#@$)) (%>))) sP_0123456789876543210))
-                       (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 10)))))
+                       (sFromInteger (sing :: Sing 10)))))
                 ((applySing
                     ((applySing ((singFun3 @(.@#@$)) (%.)))
                        ((applySing ((singFun2 @ShowStringSym0) sShowString))
                           (sing :: Sing "Just "))))
                    ((applySing
                        ((applySing ((singFun3 @ShowsPrecSym0) sShowsPrec))
-                          (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 11))))
+                          (sFromInteger (sing :: Sing 11))))
                       sArg_0123456789876543210))))
             sA_0123456789876543210
     instance SDecide a => SDecide (Maybe a) where

--- a/tests/compile-and-dump/Singletons/Nat.golden
+++ b/tests/compile-and-dump/Singletons/Nat.golden
@@ -109,7 +109,7 @@ Singletons/Nat.hs:(0,0)-(0,0): Splicing declarations
                                           -> Nat -> GHC.Types.Symbol -> GHC.Types.Symbol
     type family ShowsPrec_0123456789876543210 a a a where
       ShowsPrec_0123456789876543210 _ Zero a_0123456789876543210 = Apply (Apply ShowStringSym0 "Zero") a_0123456789876543210
-      ShowsPrec_0123456789876543210 p_0123456789876543210 (Succ arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (Data.Singletons.Prelude.Num.FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "Succ ")) (Apply (Apply ShowsPrecSym0 (Data.Singletons.Prelude.Num.FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
+      ShowsPrec_0123456789876543210 p_0123456789876543210 (Succ arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "Succ ")) (Apply (Apply ShowsPrecSym0 (FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
     type ShowsPrec_0123456789876543210Sym0 :: (~>) GHC.Types.Nat ((~>) Nat ((~>) GHC.Types.Symbol GHC.Types.Symbol))
     data ShowsPrec_0123456789876543210Sym0 a0123456789876543210
       where
@@ -246,14 +246,14 @@ Singletons/Nat.hs:(0,0)-(0,0): Splicing declarations
                  ((applySing ((singFun3 @ShowParenSym0) sShowParen))
                     ((applySing
                         ((applySing ((singFun2 @(>@#@$)) (%>))) sP_0123456789876543210))
-                       (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 10)))))
+                       (sFromInteger (sing :: Sing 10)))))
                 ((applySing
                     ((applySing ((singFun3 @(.@#@$)) (%.)))
                        ((applySing ((singFun2 @ShowStringSym0) sShowString))
                           (sing :: Sing "Succ "))))
                    ((applySing
                        ((applySing ((singFun3 @ShowsPrecSym0) sShowsPrec))
-                          (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 11))))
+                          (sFromInteger (sing :: Sing 11))))
                       sArg_0123456789876543210))))
             sA_0123456789876543210
     instance SOrd Nat => SOrd Nat where
@@ -269,7 +269,7 @@ Singletons/Nat.hs:(0,0)-(0,0): Splicing declarations
                  ((applySing ((singFun3 @FoldlSym0) sFoldl))
                     ((singFun2 @ThenCmpSym0) sThenCmp)))
                 SEQ))
-            Data.Singletons.Prelude.Instances.SNil
+            SNil
       sCompare
         (SSucc (sA_0123456789876543210 :: Sing a_0123456789876543210))
         (SSucc (sB_0123456789876543210 :: Sing b_0123456789876543210))
@@ -279,14 +279,12 @@ Singletons/Nat.hs:(0,0)-(0,0): Splicing declarations
                     ((singFun2 @ThenCmpSym0) sThenCmp)))
                 SEQ))
             ((applySing
-                ((applySing
-                    ((singFun2 @(:@#@$))
-                       Data.Singletons.Prelude.Instances.SCons))
+                ((applySing ((singFun2 @(:@#@$)) SCons))
                    ((applySing
                        ((applySing ((singFun2 @CompareSym0) sCompare))
                           sA_0123456789876543210))
                       sB_0123456789876543210)))
-               Data.Singletons.Prelude.Instances.SNil)
+               SNil)
       sCompare SZero (SSucc _) = SLT
       sCompare (SSucc _) SZero = SGT
     instance SDecide Nat => SDecide Nat where

--- a/tests/compile-and-dump/Singletons/NegativeLiterals.golden
+++ b/tests/compile-and-dump/Singletons/NegativeLiterals.golden
@@ -1,0 +1,15 @@
+Singletons/NegativeLiterals.hs:(0,0)-(0,0): Splicing declarations
+    singletons
+      [d| f :: Nat
+          f = -1 |]
+  ======>
+    f :: Nat
+    f = (-1)
+    type FSym0 :: Nat
+    type family FSym0 where
+      FSym0 = F
+    type F :: Nat
+    type family F where
+      F = Negate (FromInteger 1)
+    sF :: Sing (FSym0 :: Nat)
+    sF = sNegate (sFromInteger (sing :: Sing 1))

--- a/tests/compile-and-dump/Singletons/NegativeLiterals.hs
+++ b/tests/compile-and-dump/Singletons/NegativeLiterals.hs
@@ -1,0 +1,10 @@
+{-# LANGUAGE NegativeLiterals #-}
+module Singletons.NegativeLiterals where
+
+import Data.Singletons.TH
+import GHC.TypeNats (Nat)
+
+$(singletons [d|
+  f :: Nat
+  f = -1
+  |])

--- a/tests/compile-and-dump/Singletons/OverloadedStrings.golden
+++ b/tests/compile-and-dump/Singletons/OverloadedStrings.golden
@@ -25,7 +25,7 @@ Singletons/OverloadedStrings.hs:(0,0)-(0,0): Splicing declarations
       SymIdSym1 a0123456789876543210 = SymId a0123456789876543210
     type Foo :: Symbol
     type family Foo where
-      Foo = Apply SymIdSym0 (Data.Singletons.Prelude.IsString.FromString "foo")
+      Foo = Apply SymIdSym0 (FromString "foo")
     type SymId :: Symbol -> Symbol
     type family SymId a where
       SymId x = x
@@ -34,7 +34,7 @@ Singletons/OverloadedStrings.hs:(0,0)-(0,0): Splicing declarations
       forall (t :: Symbol). Sing t -> Sing (Apply SymIdSym0 t :: Symbol)
     sFoo
       = (applySing ((singFun1 @SymIdSym0) sSymId))
-          (Data.Singletons.Prelude.IsString.sFromString (sing :: Sing "foo"))
+          (sFromString (sing :: Sing "foo"))
     sSymId (sX :: Sing x) = sX
     instance SingI (SymIdSym0 :: (~>) Symbol Symbol) where
       sing = (singFun1 @SymIdSym0) sSymId

--- a/tests/compile-and-dump/Singletons/T166.golden
+++ b/tests/compile-and-dump/Singletons/T166.golden
@@ -45,7 +45,7 @@ Singletons/T166.hs:(0,0)-(0,0): Splicing declarations
     type family FooSym1 a0123456789876543210 where
       FooSym1 a0123456789876543210 = Foo a0123456789876543210
     type family Lambda_0123456789876543210 x s where
-      Lambda_0123456789876543210 x s = Apply (Apply (Apply FoosPrecSym0 (Data.Singletons.Prelude.Num.FromInteger 0)) x) s
+      Lambda_0123456789876543210 x s = Apply (Apply (Apply FoosPrecSym0 (FromInteger 0)) x) s
     data Lambda_0123456789876543210Sym0 x0123456789876543210
       where
         Lambda_0123456789876543210Sym0KindInference :: SameKind (Apply Lambda_0123456789876543210Sym0 arg) (Lambda_0123456789876543210Sym1 arg) =>
@@ -104,7 +104,7 @@ Singletons/T166.hs:(0,0)-(0,0): Splicing declarations
                       -> (applySing
                             ((applySing
                                 ((applySing ((singFun3 @FoosPrecSym0) sFoosPrec))
-                                   (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 0))))
+                                   (sFromInteger (sing :: Sing 0))))
                                sX))
                            sS })
     instance SFoo a =>

--- a/tests/compile-and-dump/Singletons/T178.golden
+++ b/tests/compile-and-dump/Singletons/T178.golden
@@ -141,7 +141,7 @@ Singletons/T178.hs:(0,0)-(0,0): Splicing declarations
     instance PShow Occ where
       type ShowsPrec a a a = Apply (Apply (Apply ShowsPrec_0123456789876543210Sym0 a) a) a
     sEmpty :: Sing (EmptySym0 :: [(Symbol, Occ)])
-    sEmpty = Data.Singletons.Prelude.Instances.SNil
+    sEmpty = SNil
     data SOcc :: Occ -> GHC.Types.Type
       where
         SStr :: SOcc (Str :: Occ)
@@ -185,21 +185,21 @@ Singletons/T178.hs:(0,0)-(0,0): Splicing declarations
                  ((applySing ((singFun3 @FoldlSym0) sFoldl))
                     ((singFun2 @ThenCmpSym0) sThenCmp)))
                 SEQ))
-            Data.Singletons.Prelude.Instances.SNil
+            SNil
       sCompare SOpt SOpt
         = (applySing
              ((applySing
                  ((applySing ((singFun3 @FoldlSym0) sFoldl))
                     ((singFun2 @ThenCmpSym0) sThenCmp)))
                 SEQ))
-            Data.Singletons.Prelude.Instances.SNil
+            SNil
       sCompare SMany SMany
         = (applySing
              ((applySing
                  ((applySing ((singFun3 @FoldlSym0) sFoldl))
                     ((singFun2 @ThenCmpSym0) sThenCmp)))
                 SEQ))
-            Data.Singletons.Prelude.Instances.SNil
+            SNil
       sCompare SStr SOpt = SLT
       sCompare SStr SMany = SLT
       sCompare SOpt SStr = SGT

--- a/tests/compile-and-dump/Singletons/T190.golden
+++ b/tests/compile-and-dump/Singletons/T190.golden
@@ -67,7 +67,7 @@ Singletons/T190.hs:0:0:: Splicing declarations
       Case_0123456789876543210 n 'False = Apply ErrorSym0 "toEnum: bad argument"
     type ToEnum_0123456789876543210 :: GHC.Types.Nat -> T
     type family ToEnum_0123456789876543210 a where
-      ToEnum_0123456789876543210 n = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 0))
+      ToEnum_0123456789876543210 n = Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 0))
     type ToEnum_0123456789876543210Sym0 :: (~>) GHC.Types.Nat T
     data ToEnum_0123456789876543210Sym0 a0123456789876543210
       where
@@ -82,7 +82,7 @@ Singletons/T190.hs:0:0:: Splicing declarations
       ToEnum_0123456789876543210Sym1 a0123456789876543210 = ToEnum_0123456789876543210 a0123456789876543210
     type FromEnum_0123456789876543210 :: T -> GHC.Types.Nat
     type family FromEnum_0123456789876543210 a where
-      FromEnum_0123456789876543210 T = Data.Singletons.Prelude.Num.FromInteger 0
+      FromEnum_0123456789876543210 T = FromInteger 0
     type FromEnum_0123456789876543210Sym0 :: (~>) T GHC.Types.Nat
     data FromEnum_0123456789876543210Sym0 a0123456789876543210
       where
@@ -179,7 +179,7 @@ Singletons/T190.hs:0:0:: Splicing declarations
                  ((applySing ((singFun3 @FoldlSym0) sFoldl))
                     ((singFun2 @ThenCmpSym0) sThenCmp)))
                 SEQ))
-            Data.Singletons.Prelude.Instances.SNil
+            SNil
     instance SEnum T where
       sToEnum ::
         forall (t :: GHC.Types.Nat).
@@ -193,15 +193,14 @@ Singletons/T190.hs:0:0:: Splicing declarations
                                                                      -> GHC.Types.Type) t)
       sToEnum (sN :: Sing n)
         = (id
-             @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (Data.Singletons.Prelude.Num.FromInteger 0)))))
+             @(Sing (Case_0123456789876543210 n (Apply (Apply (==@#@$) n) (FromInteger 0)))))
             (case
                  (applySing ((applySing ((singFun2 @(==@#@$)) (%==))) sN))
-                   (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 0))
+                   (sFromInteger (sing :: Sing 0))
              of
                STrue -> ST
                SFalse -> sError (sing :: Sing "toEnum: bad argument"))
-      sFromEnum ST
-        = Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 0)
+      sFromEnum ST = sFromInteger (sing :: Sing 0)
     instance SBounded T where
       sMinBound :: Sing (MinBoundSym0 :: T)
       sMaxBound :: Sing (MaxBoundSym0 :: T)

--- a/tests/compile-and-dump/Singletons/T271.golden
+++ b/tests/compile-and-dump/Singletons/T271.golden
@@ -206,14 +206,12 @@ Singletons/T271.hs:(0,0)-(0,0): Splicing declarations
                     ((singFun2 @ThenCmpSym0) sThenCmp)))
                 SEQ))
             ((applySing
-                ((applySing
-                    ((singFun2 @(:@#@$))
-                       Data.Singletons.Prelude.Instances.SCons))
+                ((applySing ((singFun2 @(:@#@$)) SCons))
                    ((applySing
                        ((applySing ((singFun2 @CompareSym0) sCompare))
                           sA_0123456789876543210))
                       sB_0123456789876543210)))
-               Data.Singletons.Prelude.Instances.SNil)
+               SNil)
     instance SEq a => SEq (Identity a) where
       (%==) ::
         forall (t1 :: Identity a) (t2 :: Identity a).
@@ -243,14 +241,12 @@ Singletons/T271.hs:(0,0)-(0,0): Splicing declarations
                     ((singFun2 @ThenCmpSym0) sThenCmp)))
                 SEQ))
             ((applySing
-                ((applySing
-                    ((singFun2 @(:@#@$))
-                       Data.Singletons.Prelude.Instances.SCons))
+                ((applySing ((singFun2 @(:@#@$)) SCons))
                    ((applySing
                        ((applySing ((singFun2 @CompareSym0) sCompare))
                           sA_0123456789876543210))
                       sB_0123456789876543210)))
-               Data.Singletons.Prelude.Instances.SNil)
+               SNil)
     instance SDecide a => SDecide (Constant a b) where
       (%~) (SConstant a) (SConstant b)
         = case ((%~) a) b of

--- a/tests/compile-and-dump/Singletons/T358.golden
+++ b/tests/compile-and-dump/Singletons/T358.golden
@@ -105,14 +105,14 @@ Singletons/T358.hs:(0,0)-(0,0): Splicing declarations
         forall b (t :: b). Sing t -> Sing (Apply Method2bSym0 t :: a)
     instance SC1 [] where
       sMethod1 :: forall a. Sing (Method1Sym0 :: [a])
-      sMethod1 = Data.Singletons.Prelude.Instances.SNil
+      sMethod1 = SNil
     instance SC2 [a] where
       sMethod2a ::
         forall b (t :: b). Sing t -> Sing (Apply Method2aSym0 t :: [a])
       sMethod2b ::
         forall b (t :: b). Sing t -> Sing (Apply Method2bSym0 t :: [a])
-      sMethod2a _ = Data.Singletons.Prelude.Instances.SNil
-      sMethod2b _ = Data.Singletons.Prelude.Instances.SNil
+      sMethod2a _ = SNil
+      sMethod2b _ = SNil
     instance SC2 a => SingI (Method2aSym0 :: (~>) b a) where
       sing = (singFun1 @Method2aSym0) sMethod2a
     instance SC2 a => SingI (Method2bSym0 :: (~>) b a) where

--- a/tests/compile-and-dump/Singletons/T371.golden
+++ b/tests/compile-and-dump/Singletons/T371.golden
@@ -45,7 +45,7 @@ Singletons/T371.hs:(0,0)-(0,0): Splicing declarations
                                           -> X a -> GHC.Types.Symbol -> GHC.Types.Symbol
     type family ShowsPrec_0123456789876543210 a a a where
       ShowsPrec_0123456789876543210 _ X1 a_0123456789876543210 = Apply (Apply ShowStringSym0 "X1") a_0123456789876543210
-      ShowsPrec_0123456789876543210 p_0123456789876543210 (X2 arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (Data.Singletons.Prelude.Num.FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "X2 ")) (Apply (Apply ShowsPrecSym0 (Data.Singletons.Prelude.Num.FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
+      ShowsPrec_0123456789876543210 p_0123456789876543210 (X2 arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "X2 ")) (Apply (Apply ShowsPrecSym0 (FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
     type ShowsPrec_0123456789876543210Sym0 :: (~>) GHC.Types.Nat ((~>) (X a) ((~>) GHC.Types.Symbol GHC.Types.Symbol))
     data ShowsPrec_0123456789876543210Sym0 a0123456789876543210
       where
@@ -85,7 +85,7 @@ Singletons/T371.hs:(0,0)-(0,0): Splicing declarations
                                           -> Y a -> GHC.Types.Symbol -> GHC.Types.Symbol
     type family ShowsPrec_0123456789876543210 a a a where
       ShowsPrec_0123456789876543210 _ Y1 a_0123456789876543210 = Apply (Apply ShowStringSym0 "Y1") a_0123456789876543210
-      ShowsPrec_0123456789876543210 p_0123456789876543210 (Y2 arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (Data.Singletons.Prelude.Num.FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "Y2 ")) (Apply (Apply ShowsPrecSym0 (Data.Singletons.Prelude.Num.FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
+      ShowsPrec_0123456789876543210 p_0123456789876543210 (Y2 arg_0123456789876543210) a_0123456789876543210 = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (>@#@$) p_0123456789876543210) (FromInteger 10))) (Apply (Apply (.@#@$) (Apply ShowStringSym0 "Y2 ")) (Apply (Apply ShowsPrecSym0 (FromInteger 11)) arg_0123456789876543210))) a_0123456789876543210
     type ShowsPrec_0123456789876543210Sym0 :: (~>) GHC.Types.Nat ((~>) (Y a) ((~>) GHC.Types.Symbol GHC.Types.Symbol))
     data ShowsPrec_0123456789876543210Sym0 a0123456789876543210
       where
@@ -174,14 +174,14 @@ Singletons/T371.hs:(0,0)-(0,0): Splicing declarations
                  ((applySing ((singFun3 @ShowParenSym0) sShowParen))
                     ((applySing
                         ((applySing ((singFun2 @(>@#@$)) (%>))) sP_0123456789876543210))
-                       (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 10)))))
+                       (sFromInteger (sing :: Sing 10)))))
                 ((applySing
                     ((applySing ((singFun3 @(.@#@$)) (%.)))
                        ((applySing ((singFun2 @ShowStringSym0) sShowString))
                           (sing :: Sing "X2 "))))
                    ((applySing
                        ((applySing ((singFun3 @ShowsPrecSym0) sShowsPrec))
-                          (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 11))))
+                          (sFromInteger (sing :: Sing 11))))
                       sArg_0123456789876543210))))
             sA_0123456789876543210
     instance SShow (X a) => SShow (Y a) where
@@ -209,14 +209,14 @@ Singletons/T371.hs:(0,0)-(0,0): Splicing declarations
                  ((applySing ((singFun3 @ShowParenSym0) sShowParen))
                     ((applySing
                         ((applySing ((singFun2 @(>@#@$)) (%>))) sP_0123456789876543210))
-                       (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 10)))))
+                       (sFromInteger (sing :: Sing 10)))))
                 ((applySing
                     ((applySing ((singFun3 @(.@#@$)) (%.)))
                        ((applySing ((singFun2 @ShowStringSym0) sShowString))
                           (sing :: Sing "Y2 "))))
                    ((applySing
                        ((applySing ((singFun3 @ShowsPrecSym0) sShowsPrec))
-                          (Data.Singletons.Prelude.Num.sFromInteger (sing :: Sing 11))))
+                          (sFromInteger (sing :: Sing 11))))
                       sArg_0123456789876543210))))
             sA_0123456789876543210
     instance Data.Singletons.ShowSing.ShowSing (Y a) =>

